# **Perform credit card fraud detection using an in-built dataset and deploy the trained model.**

In [1]:
import pandas as pd
from sklearn.datasets import fetch_openml

# Load the credit card fraud detection dataset
creditcard_data = fetch_openml(name='creditcard', version=1, as_frame=True)
df = creditcard_data.frame

# Display the first few rows of the dataset
display(df.head())

V1        V2        V3        V4        V5        V6        V7  \
0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10  ...       V21       V22       V23       V24  \
0  0.098698  0.363787  0.090794  ... -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425 -0.166974  ... -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  0.207643  ...  0.247998  0.771679  0.909412 -0.689281   
3  0.377436 -1.387024 -0.054952  ... -0.108300  0.005274 -0.190321 -1.175575   
4 -0.270533  0.817739  0.753074  ... -0.009431  0.798278 -0.137458  0.141267   

        V25       V26       V27       V28  Amount  Class  
0  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1  0.167170  0.125895 -0.008983  0.014724    2.69      0  
2 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4 -0.206010  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 30 columns]

## Explore the data

### Subtask:
Analyze the dataset to understand its structure, features, and target variable. Handle missing values and outliers if necessary.


In [2]:
# Display basic information about the DataFrame
df.info()

# Generate descriptive statistics for numerical columns
display(df.describe())

# Check for missing values
display(df.isnull().sum())

# Examine the distribution of the target variable 'Class'
display(df['Class'].value_counts())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 30 columns):
 #   Column  Non-Null Count   Dtype   
---  ------  --------------   -----   
 0   V1      284807 non-null  float64 
 1   V2      284807 non-null  float64 
 2   V3      284807 non-null  float64 
 3   V4      284807 non-null  float64 
 4   V5      284807 non-null  float64 
 5   V6      284807 non-null  float64 
 6   V7      284807 non-null  float64 
 7   V8      284807 non-null  float64 
 8   V9      284807 non-null  float64 
 9   V10     284807 non-null  float64 
 10  V11     284807 non-null  float64 
 11  V12     284807 non-null  float64 
 12  V13     284807 non-null  float64 
 13  V14     284807 non-null  float64 
 14  V15     284807 non-null  float64 
 15  V16     284807 non-null  float64 
 16  V17     284807 non-null  float64 
 17  V18     284807 non-null  float64 
 18  V19     284807 non-null  float64 
 19  V20     284807 non-null  float64 
 20  V21     284807 non-null  f

V1            V2            V3            V4            V5  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15  9.604066e-16   
std    1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00  1.380247e+00   
min   -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00 -1.137433e+02   
25%   -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01 -6.915971e-01   
50%    1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02 -5.433583e-02   
75%    1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01  6.119264e-01   
max    2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01  3.480167e+01   

                 V6            V7            V8            V9           V10  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15  2.239053e-15   
std    1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00  1.088850e+00   
min   -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01 -2.458826e+01   
25%   -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01 -5.354257e-01   
50%   -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02 -9.291738e-02   
75%    3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01  4.539234e-01   
max    7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01  2.374514e+01   

       ...           V20           V21           V22           V23  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  6.406204e-16  1.654067e-16 -3.568593e-16  2.578648e-16   
std    ...  7.709250e-01  7.345240e-01  7.257016e-01  6.244603e-01   
min    ... -5.449772e+01 -3.483038e+01 -1.093314e+01 -4.480774e+01   
25%    ... -2.117214e-01 -2.283949e-01 -5.423504e-01 -1.618463e-01   
50%    ... -6.248109e-02 -2.945017e-02  6.781943e-03 -1.119293e-02   
75%    ...  1.330408e-01  1.863772e-01  5.285536e-01  1.476421e-01   
max    ...  3.942090e+01  2.720284e+01  1.050309e+01  2.252841e+01   

                V24           V25           V26           V27           V28  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   4.473266e-15  5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16   
std    6.056471e-01  5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01   
min   -2.836627e+00 -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01   
25%   -3.545861e-01 -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02   
50%    4.097606e-02  1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02   
75%    4.395266e-01  3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02   
max    4.584549e+00  7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   

              Amount  
count  284807.000000  
mean       88.349619  
std       250.120109  
min         0.000000  
25%         5.600000  
50%        22.000000  
75%        77.165000  
max     25691.160000  

[8 rows x 29 columns]

V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

Class
0    284315
1       492
Name: count, dtype: int64

## Preprocess the data

### Subtask:
Prepare the data for model training by scaling features, handling imbalanced classes, and splitting the data into training and testing sets.


In [3]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from imblearn.over_sampling import SMOTE

# Separate features (X) and target (y)
X = df.drop('Class', axis=1)
y = df['Class']

# Scale the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Split data into training and testing sets with stratification
# Keep the test set aside before applying SMOTE
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42, stratify=y)

# Apply SMOTE only to the training data
smote = SMOTE(random_state=42)
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

# Display the shape of the original and resampled training data
print("Original training data shape:", X_train.shape)
print("Resampled training data shape:", X_train_resampled.shape)
print("Original training target distribution:\n", y_train.value_counts())
print("Resampled training target distribution:\n", y_train_resampled.value_counts())
print("Test target distribution:\n", y_test.value_counts())

Original training data shape: (227845, 29)
Resampled training data shape: (454902, 29)
Original training target distribution:
 Class
0    227451
1       394
Name: count, dtype: int64
Resampled training target distribution:
 Class
0    227451
1    227451
Name: count, dtype: int64
Test target distribution:
 Class
0    56864
1       98
Name: count, dtype: int64


## Train a model

### Subtask:
Choose a suitable model for fraud detection (e.g., Logistic Regression, Random Forest, or a neural network) and train it on the training data.


In [4]:
from sklearn.linear_model import LogisticRegression

# Instantiate a LogisticRegression model
model = LogisticRegression(solver='liblinear', random_state=42)

# Train the model using the resampled training data
model.fit(X_train_resampled, y_train_resampled)

LogisticRegression(random_state=42, solver='liblinear')

## Evaluate the model

### Subtask:
Evaluate the model's performance using appropriate metrics (e.g., precision, recall, F1-score, AUC) on the testing data.


In [5]:
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score

# Predict classes on the test set
y_pred = model.predict(X_test)

# Predict probabilities for the positive class on the test set
y_pred_proba = model.predict_proba(X_test)[:, 1]

# Generate and print the classification report
print("Classification Report:")
print(classification_report(y_test, y_pred))

# Generate and print the confusion matrix
print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

# Calculate and print the ROC AUC score
roc_auc = roc_auc_score(y_test, y_pred_proba)
print(f"ROC AUC Score: {roc_auc}")

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.97      0.99     56864
           1       0.06      0.92      0.11        98

    accuracy                           0.97     56962
   macro avg       0.53      0.95      0.55     56962
weighted avg       1.00      0.97      0.98     56962

Confusion Matrix:
[[55341  1523]
 [    8    90]]
ROC AUC Score: 0.9706652392245587


## Summary:

### Data Analysis Key Findings

*   The dataset contains 284,807 entries and 30 columns, with no missing values.
*   There is a significant class imbalance in the target variable 'Class', with 284,315 non-fraudulent transactions (Class 0) and only 492 fraudulent transactions (Class 1).
*   Feature scaling was applied to the dataset.
*   SMOTE was applied to the training data to address the class imbalance, resulting in a balanced distribution of 227,452 instances for both classes in the resampled training set.
*   A Logistic Regression model was trained on the resampled training data.
*   The model achieved a high recall of 0.92 for the fraudulent class (class 1) on the test set, indicating it identifies most fraudulent transactions.
*   The precision for the fraudulent class (class 1) is low at 0.06, suggesting a high number of false positives.
*   The confusion matrix shows 55341 true negatives, 90 true positives, 1523 false positives, and 8 false negatives.
*   The ROC AUC score is 0.9707, indicating good overall discriminative ability.
*   The trained Logistic Regression model was successfully saved to a file named `logistic_regression_model.joblib`.

### Insights or Next Steps

*   While the model has high recall for fraud detection, the low precision indicates a need to reduce false positives. Further model tuning or exploring different algorithms (e.g., tree-based models, anomaly detection) could help improve precision while maintaining high recall.
*   The saved model can now be loaded and integrated into an application or system to perform real-time or batch predictions on new credit card transactions to detect potential fraud.
